Don't derive type var constraints from type param bounds

adriaanm opened the pull request:
This overly constrains type checking, as reported in #6739 and scala/bug#10819.

dwijnand commented:
Looks like you need to be less ambitious. Or differently ambitious.

```
[error] /home/travis/build/scala/scala/src/library/scala/collection/concurrent/TrieMap.scala:986: type mismatch;
[error]  found   : scala.collection.concurrent.TrieMap[K,V]
[error]  required: scala.collection.concurrent.TrieMap[K,V] with scala.collection.mutable.Growable[Elem]
[error]   def newBuilder[K, V] = new GrowableBuilder(empty[K, V])
[error]                                                   ^
```

(one of a number of failures in Travis CI)

adriaanm commented:
As usual, my ambitions are being adjusted towards the more ad-hoc. I think I have a compromise brewing.

adriaanm commented:
Bah, it passes the test suite, though.

adriaanm commented:
As usual, too much code has started to rely on this. Not sure we can do much here.

adriaanm commented:
No, we really can't let this escape into the wild. It's only been enabled since M3, so not sure why so many tests fail. Must be doing something else wrong.

adriaanm committed:
Don't propagate type param bounds to TypeVar

This leads to overfitting during type checking,
because you'll infer a type based solely on the type params'
bounds, and then take that as the expected type down the road,
even though the later type may actually be more specific.

adriaanm commented:
/cc @milessabin

milessabin commented:
@adriaanm that looks a lot nicer.

> TODO: why do we need the additional hi-bound? see sip23-widen

What's the issue with sip23-widen?

milessabin commented:
@adriaanm Ahh ... OK, I see. We need to propagate the `<: Singleton` from the symbol to the constraint so that it's available [here](https://github.com/scala/scala/blob/2.13.x/src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala#L258). Just stopping widening isn't sufficient.

I'm not saying there isn't a better way to do that ...

milessabin commented:
Hmm ... or perhaps not. Something must be being propagated from the constraint `hiBounds` to some other type variable somewhere.

adriaanm commented:
Ok, thanks for checking & clarifying. I think that makes sense then.

lrytz commented on test/files/run/t10819.scala:
```diff
@@ -0,0 +1,23 @@
+
+    val inferredType = out.lines.filter(_.contains("C.this.id[")).map(_.trim).toList
+    assert(inferredType.length == 1)
+    assert(inferredType.forall(_.startsWith("C.this.id[Int]")), inferredType)
```
could also be a ReplTest, the transcript in the check file would show the inferred type

lrytz commented on test/files/pos/S1.scala:
```diff
@@ -8,24 +8,6 @@
-class S2() {
-  def foo[T <: this.type](x: T) = x
-  def f = foo(this)
-}
```
Why was this removed? It still compiles after this PR.

It would also be a good opportunity to rename the file of this test :-) And the comment in the beginning looks stale.

adriaanm replied:
Hmm, the test failed on my machine, but maybe that was spurious.

lrytz replied:
I tried with `scala-pr 6789` in the repl.

adriaanm replied:
Maybe I had a stray class file or something. Pushed a revert for it

lrytz reviewed (commented):
Change LGTM

adriaanm committed:
Restore test/files/pos/S1.scala
